### PR TITLE
UMI-workflow using Sentieon tools

### PR DESCRIPTION
**Pull request by ashwini06**
### This PR:
Workflow to perform UMI tag extraction using Sentieon tools and calling somatic variants 

**Features**
- [x] Independent snakemake workflow for running UMI-samples   
- [x] Validated tests on internal projects (Dilution series, SeraCare Samples) 
- [x] Calling SNVs - TNscope, Vardict, mutect2
- [x] VEP annotation 
- [x] Generate relevant AF tables for plotting 
- [x] Integrate the workflow into BALSAMIC 

UMI workflow can be run in the BALSAMIC workflow by specifying the analysis-type to '**umi'**

s1

Steps to run the UMI workflow 

###### New feature:
   --background-variants [ Optional file] -  required to create table with list of variants and their Allelic frequencies

```
balsamic config case --tumor /home/proj/long-term-stage/cancer/UMI_analysis_runs_APJ/ACC5611A1-A9/fairsalmon/fastq/concatenated_ACC5611A1_XXXXXX_R_1.fastq.gz --case-id fairsalmon2 --analysis-dir demo/ --panel-bed /home/proj/long-term-stage/cancer/UMI_analysis_runs_APJ/SeraCare_VCFs/gicfdna_3.1_hg19_design.bed --background-variants /home/proj/long-term-stage/cancer/UMI_analysis_runs_APJ/SeraCare_VCFs/SeraCare_16variants_pos.txt --reference-config BALSAMIC_reference/GRCh37/reference.json --singularity BALSAMIC/containers/BALSAMIC_latest.sif
```

a1

 
##### Run the UMI-workflow

```
balsamic run analysis -a umi --sample-config demo/fairsalmon2/fairsalmon2.json --run-mode local --snakemake-opt "--cores 8"  --run-analysis 
```

a2

### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

**Comment by ashwini06** [comment]
UMI workflow can be run in the BALSAMIC workflow by specifying the analysis-type to 'umi'

s1

Command-line 

```
$balsamic run analysis --sample-config demo/demo_run_balsamic/demo_run_balsamic.json -a umi
```
s2

File: BALSAMIC/snakemake_rules/umi/annotate_vep.rule [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# vim: syntax=python tabstop=4 expandtab
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/umi/annotate_vep.rule:1]: Can we use BALSAMIC/snakemake_rules/annotation/vep.rule instead?

File: BALSAMIC/snakemake_rules/umi/sentieon_consensuscall.rule [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# vim: syntax=python tabstop=4 expandtab
+# coding: utf-8
+
+shell.prefix('set -eo pipefail;')
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/umi/sentieon_consensuscall.rule:4]: might be better to move this to Snakefile instead.

File: BALSAMIC/snakemake_rules/umi/sentieon_umiextract.rule [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# vim: syntax=python tabstop=4 expandtab
+# coding: utf-8
+
+shell.prefix('set -eo pipefail;')
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/umi/sentieon_umiextract.rule:4]: same as above. move it to Snakefile instead.

File: BALSAMIC/snakemake_rules/umi/sentieon_varcall_tnscope.rule [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# vim: syntax=python tabstop=4 expandtab
+# coding: utf-8
+
+shell.prefix('set -eo pipefail;')
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/umi/sentieon_varcall_tnscope.rule:4]: ditto

File: BALSAMIC/snakemake_rules/umi/varcall_vardict.rule [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# vim: syntax=python tabstop=4 expandtab
+# coding: utf-8
+
+shell.prefix('set -eo pipefail;')
```
Comment (hassanfa) [inline BALSAMIC/snakemake_rules/umi/varcall_vardict.rule:4]: here as well

**Review by hassanfa — APPROVED** [review]
Great work 🌟 🌟 🌟  Just some minor comments on rules with shell.prefix(..).

Merge whenever you're happy with the result.

File: BALSAMIC/utils/models.py [hunk 6]
```diff
@@ -332,6 +332,7 @@ class BalsamicConfigModel(BaseModel):
     samples: Dict[str, SampleInstanceModel]
     reference: Dict[str, Path]
     singularity: FilePath
+    background_variants: FilePath
```
Comment (Mropat) [inline BALSAMIC/utils/models.py:335]: Are we always expecting to set this field? if not, it should be optional

File: BALSAMIC/snakemake_rules/umi/annotate_vep.rule (hunk 1 above)
Comment (ashwini06) [reply BALSAMIC/snakemake_rules/umi/annotate_vep.rule:1]: ` BALSAMIC/snakemake_rules/annotation/vep.rule` contains `{case_name}` as wildcards whereas UMI workflow rules are built based on wildcard `{sample}`. We can mimic the `{sample}` to `{case_name}` in order to just reuse BALSAMIC vep rule. But the final output filenames are generated with `{sample}` names only. For now, updated params as according to BALSAMIC vep rule.

**Comment by ashwini06** [comment]
> Great work 🌟 🌟 🌟 Just some minor comments on rules with shell.prefix(..).
> 
> Merge whenever you're happy with the result.

Thanks Hassan.

**Pull request merged**